### ISSUE TITLE
How to handle re-zipping of files

### ISSUE THREAD

**Issue by georgemccabe** [NCAR/METplus#89]
We would like a conf option to rezip file that were unzipped after processing. There are many things we need to discuss to figure out how we want to implement this functionality.

- How can we keep track of which type of zipping was used to make sure we re-zip the same way?

- Do we want a utility to run at the end of the PROCESS_LIST to re-zip so that there isn't redundant zipping/unzipping if multiple apps use the same files? We could keep track of a list of files that were unzipped and their zipping type (unique list) then re-zip all of the items in that list.

We need to discuss how we are going to handle this further before we implement it. Please comment on this issue with any ideas, suggestions, comments, etc.

**Comment by georgemccabe** [comment]
Some other things to think about: 

- Some of the unzipping programs left the original zipped file in the directory. This could cause issues trying to create a new zip file that already exists

- Users running MET+ may not have permission to write files to the input data directory

- In general, we should preserve the input data as much as possible. Messing with zipping/unzipping data in these areas are risky and could potentially wipe out user data. This cannot happen. We discussed the possibility of creating a staging or temp directory where we copy files to unzip them to make sure that we don't mess up input data. We discussed using the process ID to create a subdirectory so that multiple instances of MET+ don't step on each others toes. This data would be removed after a run of MET+. We need to figure out how to maintain references to these files, i.e. if one file read by MET+ is unzipped so the path is the input directory while another file needs to be unzipped so it is copied and lives in the staging/temp directory. 
    - should we copy all input data into a staging area? how will MET+ know this is an input that must be preserved? (first item in PROCESS_LIST?)

These and other related need to be discussed in detail before we can move forward with implementing this behavior.

**Comment by georgemccabe** [comment]
A branch called feature_89_unzip has modifications to preprocess files with .gz, .bz2, .zip, and .grd extensions. The first 3 are unzipped with the proper utility, while .grd files are passed to GempakToCF. All files are placed in a staging directory (default below OUTPUT_BASE) and the path to the new file is passed to the wrappers if successful. Wrappers can utilize this functionality be called util.preprocess_file on the file. The function returns None if it is unable to process the data, so the wrapper should always check if the result is not equal to None before using the output.

For now the extensions listed above are required for METplus to know how to handle the files. We cannot rely on this, so we will need to use a utility to identify the file type. The following comment contains information on a python utility called python-magic that could potentially be used to solve this issue.

**Comment by georgemccabe** [comment]
I installed the python-magic library on my personal computer to test out the capabilities for examining file types in python. It was easy to install via pip. There are a few extra steps when installing on Windows or OSX that are described in the GitHub readme which were also pretty straight forward for me.

https://github.com/ahupp/python-magic

It looks like you are able to use the utility to get the type for grib, grib2, netcdf, bzip2, zip, and gzip. Gempak files just reported 'data.' However, I did notice that when I list the contents of a gempak file with less, the first characters are 'GEMPAK DATA MANAGEMENT FILE' This could be used to identify the file type of these files. We also need to be aware that the 'file' unix command has output HDF as the file type for some NetCDF files, so we will need to handle this case appropriately.

>>> magic.from_file("/Users/funkle/Work/data/StageIV_Gempak/2017050912_st4.grd")
'data'
>>> magic.from_file("/Users/funkle/Work/data/PHPT/20170622/20170622_i06_f006_HRRRTLE_PHPT.grb2")
'Gridded binary (GRIB) version 2'
>>> magic.from_file("/Users/funkle/Work/data/sample_fcst/2005080700/wrfprs_ruc13_00.tm00_G212")
'Gridded binary (GRIB) version 1'
>>> magic.from_file("/Users/funkle/Work/data/uswrp/NationalBlend/native/20170621/blendp_qpf_2017062100f024.nc")
'NetCDF Data Format data'
>>> magic.from_file("/Users/funkle/Work/data/zip/testfile2.txt.bz2")
'bzip2 compressed data, block size = 900k'
>>> magic.from_file("/Users/funkle/Work/data/zip/testfile3.txt.zip")
'Zip archive data, at least v1.0 to extract'
>>> magic.from_file("/Users/funkle/Work/data/zip/testfile4.txt")
'empty'
>>> magic.from_file("/Users/funkle/Work/data/zip/testfile.txt.gz")
'gzip compressed data, was "testfile.txt", last modified: Wed Nov 14 22:11:22 2018, from Unix'

**Comment by georgemccabe** [comment]
The changes from this issue were merged into the develop branch. The utility was renamed to preprocess_file from decompress_file. It takes a filename, p conf object, and an option logger object. I am leaving this ticket open because the modifications involving processing based on the actual file type instead of the extension are related.

**Comment by georgemccabe** [comment]
This issue has been resolved by using a staging dir. However, we can improve upon the unzipping capabilities, by checking for other compressed file naming formats (for example, BZIP2 instead of bzip2), or even checking the files to see what type of file format they are.
